### PR TITLE
Draw to insert

### PR DESCRIPTION
**Pull request by gbalint**
# Draw to insert feature

## Superstrategy

To implement draw to insert I created a new superstrategy, which does the following

1. inserts the element from `canvasState.interactionTarget`
2. finds the best strategy to handle the drawing interaction, with a few modifications:
- the drag interaction is with zero drag vector
- with cmd down
- with the new inserted element as target
This strategy will be a reparent to the starting point of the drawn rectangle
3. finds the best strategy to handle the drag movement:
- the drag handle should be a resize handle
- with the new inserted (and reparented) element as target

The upside of this solution is that it fully reuses all the intricacies of reparenting and resizing (e.g. how to reparent into a flex container). The downside is that I really had to mock the interactions and the metadata to make sure the normal dragging strategies are "tricked" into working after a fresh insertion. The biggest issue was that I needed metadata for the new inserted element, and I solved that by creating a fake empty metadata which just includes the necessary stuff to make the dragging strategies work well.

## Flex resize basic

I added a very simple flex resize strategy, because I needed to make resize work in flex containers (because the superstrategy relies on resizing)

## Insert mode changes

In insert-mode-control-container I start a resize drag interaction with an insertion subject as `interactionTarget`

# TODO

- [ ] The recently implemented click to insert is broken, fix it
- [ ] insert-mode-control-container contains special logic for images, it seems to be broken, but still, let's check if anything needs to ported from there
- [ ] remove unused code from insert-mode-control-container
- [ ] find better way to force flex reparent to render the element and not just position indicators (currently I trick it with saying it is an `end-interaction`)

### CODE REVIEW

File: editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.tsx [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    {
+      left: frame.x,
+      top: frame.y,
+      width: frame.width,
+      height: frame.height,
+    },
+    ['style'],
+  )
+
+  if (isLeft(updatedAttributes)) {
```
Comment (seanparsons) [inline editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.tsx:194]: Maybe a bit cleaner with a `foldEither`.

File: editor/src/components/canvas/canvas-strategies/flex-resize-basic-strategy.tsx [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+      width: newWidth,
+      height: newHeight,
+    })
+  } else {
+    const isEdgeHorizontalSide = isEdgePositionAHorizontalEdge(edgePosition)
+
+    const oppositeSideCenterPosition = isEdgeHorizontalSide
+      ? ({
+          x: edgePosition.x,
+          y: 1 - edgePosition.y,
```
Comment (seanparsons) [inline editor/src/components/canvas/canvas-strategies/flex-resize-basic-strategy.tsx:162]: There's some utility functions like `oppositeEdgePositionPart`, which I _believe_ means you can avoid those casts.

**Review by seanparsons — APPROVED** [review]

**Review by enidemi — APPROVED** [review]

File: editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.tsx (hunk 1 above)
Comment (gbalint) [reply editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.tsx]: Fixed in d22fe17f8bdd0ad0bb7d8bc117885150d0f44a25

File: editor/src/components/canvas/canvas-strategies/flex-resize-basic-strategy.tsx (hunk 2 above)
Comment (gbalint) [reply editor/src/components/canvas/canvas-strategies/flex-resize-basic-strategy.tsx]: Good idea, fixed in f676396b2bef16107049579f9f56fbec7938aed5

File: editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx [hunk 3]
```diff
@@ -0,0 +1,766 @@
...
+} from '../ui-jsx.test-utils'
+import { CanvasControlsContainerID } from '../controls/new-canvas-controls'
+import { wait } from '../../../utils/utils.test-utils'
+
+function slightlyOffsetWindowPointBecauseVeryWeirdIssue(point: {
+  x: number
+  y: number
+}): WindowPoint {
+  // FIXME when running in headless chrome, the result of getBoundingClientRect will be slightly
+  // offset for some unknown reason, meaning the inserted element will be 1 pixel of in each dimension
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx:19]: 👀

Comment (gbalint) [reply editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx:19]: I haven't touched this, this is just renaming of insert-mode.spec.browser2.tsx and I added new test cases

File: editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+    })
+    const endPoint = slightlyOffsetWindowPointBecauseVeryWeirdIssue({
+      x: targetElementBounds.x + 25,
+      y: targetElementBounds.y + 305,
+    })
+
+    // Drag from inside bbb to inside ccc
+    await fireDragEvent(canvasControlsLayer, startPoint, endPoint)
+
+    // Check that the inserted element is a child of bbb
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx:342]: the code instead is checking that the inserted element is a sibling of bbb, someone is not telling the truth here!

File: editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx [hunk 5]
```diff
@@ -0,0 +1,766 @@
...
+    })
+    const endPoint = slightlyOffsetWindowPointBecauseVeryWeirdIssue({
+      x: targetElementBounds.x + targetElementBounds.width + 25,
+      y: targetElementBounds.y + targetElementBounds.height + 305,
+    })
+
+    // Drag from inside bbb to inside ccc
+    await fireDragEvent(canvasControlsLayer, startPoint, endPoint)
+
+    // Check that the inserted element is a child of bbb
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx:406]: the inserted element here is a sibling _after_ bbb!

File: editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx [hunk 6]
```diff
@@ -0,0 +1,766 @@
...
+    })
+    const endPoint = slightlyOffsetWindowPointBecauseVeryWeirdIssue({
+      x: targetElementBounds.x + targetElementBounds.width + 5,
+      y: targetElementBounds.y + 5,
+    })
+
+    // Drag from inside bbb to inside ccc
+    await fireDragEvent(canvasControlsLayer, startPoint, endPoint)
+
+    // Check that the inserted element is a child of bbb
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx:470]: the inserted element is a sibling after bbb!

Comment (gbalint) [reply editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx:470]: Oh sorry, copy pasta comment

File: editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx (hunk 5 above)
Comment (gbalint) [reply editor/src/components/canvas/canvas-strategies/draw-to-insert-strategy.spec.browser2.tsx:406]: Oh sorry, copy pasta comment

**Review by balazsbajorics — APPROVED** [review]

**Pull request merged**